From 2eaf28eccae97857d712cfa613cec2bebbb6d802 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Caol=C3=A1n=20McNamara?= <caolanm@redhat.com>
Date: Sat, 6 Nov 2021 21:32:21 +0000
Subject: [PATCH] Resolves: tdf#145567 restore focus to the usual frame focus
 widget

when tearing down the start center. Don't leave the focus in an
arbitrary widget.

Change-Id: I82c30c94121dc43b2ea1b4fbc66a0a3e79f7e664
---
 vcl/unx/gtk3/gtk3gtkinst.cxx | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/vcl/unx/gtk3/gtk3gtkinst.cxx b/vcl/unx/gtk3/gtk3gtkinst.cxx
index a5d233e2f95e..fce7afe90ef6 100644
--- a/vcl/unx/gtk3/gtk3gtkinst.cxx
+++ b/vcl/unx/gtk3/gtk3gtkinst.cxx
@@ -16635,6 +16635,14 @@ private:
         // rehook handler and let vcl cycle its own way through this widget's
         // children
         pFrame->AllowCycleFocusOut();
+
+        // tdf#145567 if the focus is in this hierarchy then, now that we are tearing down,
+        // move focus to the usual focus candidate for the frame
+        GtkWindow* pFocusWin = get_active_window();
+        GtkWidget* pFocus = pFocusWin ? gtk_window_get_focus(pFocusWin) : nullptr;
+        bool bHasFocus = pFocus && gtk_widget_is_ancestor(pFocus, pTopLevel);
+        if (bHasFocus)
+            pFrame->GrabFocus();
     }
 
     static void signalUnmap(GtkWidget*, gpointer user_data)
-- 
2.33.1

